### PR TITLE
soc: sam0: add support for setting low power states

### PR DESCRIPTION
**Pull request by benpicco**
This adds support for the three Idle modes and the Deep Sleep mode on the Atmel SAM0 SoCs.

I originally included this in #13494 but it makes more sense to keep this a separate pull request.

### CODE REVIEW

File: soc/arm/atmel_sam0/common/power.c [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+static inline void sam0_deep_sleep(void)
+{
+	SCB->SCR |= SCB_SCR_SLEEPDEEP_Msk;
+	__DSB();
+	__WFI();
+}
+
+/* Invoke Low Power/System Off specific Tasks */
+void sys_set_power_state(enum power_states state)
+{
```
Comment (Sizurka) [inline soc/arm/atmel_sam0/common/power.c]: Nit: Errata 1.14.2 on SAMD21 (wake failures) silicon revision <= C (DSU->DID.bit.REVISION <= 0x2) should probably be accounted for here (i.e. write `NVMCTRL->CTRLB.bit.SLEEPPRM = NVMCTRL_CTRLB_SLEEPPRM_DISABLED_Val`).  I have no idea about if those are actually in the wild or just on paper in the errata, so I don't know if that's really required.

**Review by Sizurka — APPROVED** [review]

**Comment by benpicco** [comment]
needs rebasing after #14448

File: soc/arm/atmel_sam0/common/power.c (hunk 1 above)
Comment (benpicco) [reply soc/arm/atmel_sam0/common/power.c]: I've added a check for this errata.

File: soc/arm/atmel_sam0/common/power.c [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+#endif
+}
+
+/* Invoke Low Power/System Off specific Tasks */
+void sys_set_power_state(enum power_states state)
+{
+	switch (state) {
+
+#ifdef CONFIG_SYS_POWER_SLEEP_STATES
+	case SYS_POWER_STATE_SLEEP_1:
```
Comment (pizi-nordic) [inline soc/arm/atmel_sam0/common/power.c]: The SYS_POWER_STATE_SLEEP_X defines might be not available if user disable some states in Kconfig
You have to wrap these in `if CONFIG_HAS_SYS_POWER_STATE_SLEEP_X`.

The same applies for deep sleep states.

File: soc/arm/atmel_sam0/common/power.c [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	default:
+		LOG_ERR("Unsupported power state %u", state);
+		break;
+	}
+}
+
+/* Handle SOC specific activity after Low Power Mode Exit */
+void sys_power_state_post_ops(enum power_states state)
+{
+	switch (state) {
```
Comment (pizi-nordic) [inline soc/arm/atmel_sam0/common/power.c]: IMHO this whole switch case might be removed.

**Review by pizi-nordic — CHANGES_REQUESTED** [review]

File: soc/arm/atmel_sam0/common/power.c (hunk 2 above)
Comment (benpicco) [reply soc/arm/atmel_sam0/common/power.c]: Ok.

**Review by ioannisg — COMMENTED** [review]
Is this abandoned @benpicco ?

**Comment by benpicco** [comment]
I’m afraid so.

**Pull request closed without merging**